node/src/components/consensus/utils.rs

```diff
@@ -52,5 +58,170 @@ pub(crate) fn check_sufficient_finality_signatures(
         });
     }
 
+    // Verify: the total weight decreased by the smallest weight of a single signature should be
+    // below threshold (anti-spam countermeasure).
+    if weight_minus_minimum * U512::from(*lower_bound.denom())
+        > total_weight * U512::from(*lower_bound.numer())
+    {
+        return Err(FinalitySignatureError::TooManySignatures {
+            trusted_validator_weights: trusted_validator_weights.clone(),
+            block_signatures: Box::new(block_signatures.clone()),
+            signature_weight: Box::new(signature_weight),
+            weight_minus_minimum: Box::new(weight_minus_minimum),
+            total_validator_weight: Box::new(total_weight),
+            finality_threshold_fraction,
+        });
+    }
+
     Ok(())
 }
+
+pub(crate) fn get_minimal_set_of_signatures(
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+    finality_threshold_fraction: Ratio<u64>,
+    mut block_signatures: BlockSignatures,
+) -> Option<BlockSignatures> {
+    // Calculate the values for comparison.
+    let total_weight: U512 = trusted_validator_weights
+        .iter()
+        .map(|(_, weight)| *weight)
+        .sum();
+
+    let lower_bound = (finality_threshold_fraction + 1) / 2;
+
+    let mut sig_weights: Vec<_> = block_signatures
+        .proofs
+        .keys()
+        .filter_map(|pub_key| {
+            trusted_validator_weights
+                .get(pub_key)
+                .map(|weight| (pub_key.clone(), *weight))
+        })
+        .collect();
+    // Sort descending by weight.
+    sig_weights.sort_unstable_by(|a, b| b.1.cmp(&a.1));
```

Comment:
We could do this whole thing in one pass by maintaining a heap of the signatures we want to keep sorted by weight.

Probably not worth it, just taking a note…